cmd/crane/rebase.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2018 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+
+	"github.com/google/go-containerregistry/authn"
+	"github.com/google/go-containerregistry/name"
+	"github.com/google/go-containerregistry/v1/mutate"
+	"github.com/google/go-containerregistry/v1/remote"
+	"github.com/spf13/cobra"
+)
+
+func init() {
+	var orig, oldBase, newBase, rebased string
+	rebaseCmd := &cobra.Command{
+		Use:   "rebase",
+		Short: "Rebase an image onto a new base image",
+		Args:  cobra.NoArgs,
+		Run: func(*cobra.Command, []string) {
+			rebase(orig, oldBase, newBase, rebased)
+		},
+	}
+	rebaseCmd.Flags().StringVarP(&orig, "original", "", "", "Original image to rebase")
+	rebaseCmd.Flags().StringVarP(&oldBase, "old_base", "", "", "Old base image to remove")
+	rebaseCmd.Flags().StringVarP(&newBase, "new_base", "", "", "New base image to insert")
+	rebaseCmd.Flags().StringVarP(&rebased, "rebased", "", "", "Tag to apply to rebased image")
+	rootCmd.AddCommand(rebaseCmd)
+}
+
+func rebase(orig, oldBase, newBase, rebased string) {
+	if orig == "" || rebased == "" {
+		log.Fatalln("Must provide --original and --rebased")
+	}
+
+	origImg, origRef, err := getImage(orig)
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	if oldBase == "" || newBase == "" {
+		// Attempt to detect rebase hints in original image.
+		hint, err := mutate.DetectRebaseHint(origImg)
+		if err != nil {
+			log.Fatalln(err)
+		}
+		if hint == nil {
+			log.Fatalln("--old_base and --new_base were not provided, and rebase hint could not be found")
+		}
+		oldBase, newBase = hint.OldBase, hint.NewBase
+	}
+
+	oldBaseImg, oldBaseRef, err := getImage(oldBase)
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	newBaseImg, newBaseRef, err := getImage(newBase)
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	rebasedTag, err := name.NewTag(rebased, name.WeakValidation)
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	var newBaseTag name.Tag
+	var newBaseDigest name.Digest
+	if t, ok := newBaseRef.(name.Tag); ok {
+		newBaseTag = t
+		dig, err := newBaseImg.Digest()
+		if err != nil {
+			log.Fatalln(err)
+		}
+		newBaseDigest, err = name.NewDigest(fmt.Sprintf("%s@%s", newBaseTag.Context(), dig), name.WeakValidation)
+		if err != nil {
+			log.Fatalln(err)
+		}
+	}
```

Comment:
I'm not sure why `mutate.Rebase` doesn't simply take a `name.Reference` and do this internally?

That feels like it would be friendlier to clients of the API.

Reply:
+1 to generic rebase :)

We might want to add something to make instantiating a remote.Image a bit easier (since you almost always want `DefaultKeychain`, etc). We could just move it from `cmd/crane/get.go` to be part of the library... `remote.ImageFromRef(ref) -> (v1.Image, error)` or something.